docs/source/quickstart.rst

```diff
@@ -0,0 +1,389 @@
+function called :func:`trio.testing.open_stream_to_socket_listener`
+that can take a :class:`~trio.SocketListener` and make a connection to
+it.
+
+Putting it all together::
+
+   from trio.testing import open_stream_to_socket_listener
+
+   # Don't copy this -- we can do better
+   async def test_attempt_3():
+       async with trio.open_nursery() as nursery:
+           # Start server running in the background
+           # AND wait for it to finish starting up before continuing
+           # AND find out where it's actually listening
+           listeners = await nursery.start(
+               partial(trio.serve_tcp, echo_server_handler, port=0)
+           )
+
+           # Connect to the server.
+           # There might be multiple listeners (example: IPv4 and
+           # IPv6), but we don't care which one we connect to, so we
+           # just use the first.
+           client = await open_stream_to_socket_listener(listeners[0])
+           async with client:
+               for test_byte in [b"a", b"c", b"c"]:
+                   await client.send_all(test_byte)
+                   assert await client.receive_some(1) == test_byte
+
+Okay, this is getting closer... but if we try to run it, we'll find
+that it just hangs instead of completing. What's going on?
+
+The problem is that after we finish doing our tests, we need to shut
+down the server – otherwise it will keep waiting for new connections
+forever, and the test will never finish. We can fix this by cancelling
+the nursery once we're done with it::
+
+   # Don't copy this -- it finally works, but we can still do better!
+   async def test_attempt_4():
+       async with trio.open_nursery() as nursery:
+           # Start server running in the background
+           # AND wait for it to finish starting up before continuing
+           # AND find out where it's actually listening
+           listeners = await nursery.start(
+               partial(trio.serve_tcp, echo_server_handler, port=0)
+           )
+
+           # Connect to the server.
+           # There might be multiple listeners (example: IPv4 and
+           # IPv6), but we don't care which one we connect to, so we
+           # just use the first.
+           client = await open_stream_to_socket_listener(listeners[0])
+           async with client:
+               for test_byte in [b"a", b"c", b"c"]:
+                   await client.send_all(test_byte)
+                   assert await client.receive_some(1) == test_byte
+
+           # Shut down the server now that we're done testing it
+           nursery.cancel_scope.cancel()
+
+Okay, finally this test works correctly. But that's a lot of
+boilerplate. Remember, we need to write lots of tests for this server,
+and we don't want to have to copy-paste all that stuff into every
+test. Let's factor it out into a fixture.
+
+Probably our first attempt will look something like::
+
+   # DON'T DO THIS, IT DOESN'T WORK
```

Comment:
Unfortunately, we're both right – it doesn't work, *and* it's extremely useful. Since the text is correct I'm going to leave it for now, and let's discuss how we can fix it over here: https://github.com/python-trio/pytest-trio/issues/55